src/tools/optimization-options.h

```diff
@@ -30,7 +30,36 @@ struct OptimizationOptions : public ToolOptions {
   static constexpr const int OS_OPTIMIZE_LEVEL = 2;
   static constexpr const int OS_SHRINK_LEVEL = 1;
 
-  std::vector<std::string> passes;
+  // Information to run a pass, as requested by a commandline flag.
+  struct PassInfo {
+    // The name of the pass to run.
+    std::string name;
+
+    // The optimize and shrink levels to run the pass with, if specified. If not
+    // specified then the defaults are used.
+    std::optional<int> optimizeLevel;
+    std::optional<int> shrinkLevel;
+
+    PassInfo(std::string name) : name(name) {}
+    PassInfo(const char* name) : name(name) {}
+    PassInfo(std::string name, int optimizeLevel, int shrinkLevel)
+      : name(name), optimizeLevel(optimizeLevel), shrinkLevel(shrinkLevel) {}
+  };
+
+  std::vector<PassInfo> passes;
+
+  // Add a request to run all the default opt passes. They are run with the
+  // current opt and shrink levels specified, which are read from passOptions.
+  //
+  // Each caller to here sets the opt and shrink levels before, which provides
+  // the right values for us to read. That is, -Os etc. sets the default opt
```

Comment:
This would be simpler if the levels were arguments to `addDefaultOptPasses`. Is there a reason not to do that?

Reply:
That would take more work, I think. That method would need to temporarily override the default opt levels like we do in this PR. But it would need to create a new pass runner to do so, I suspect. At minimum it would be a similar amount of code, if not more. And definitely more if we also need such a capability in a general pass and not just the default opt passes, which I think is possible in the future (then all `add*()` methods would need to get those arguments).